File: 001-counter/program/src/instruction.rs
use borsh::{BorshDeserialize, BorshSerialize};

#[derive(BorshSerialize, BorshDeserialize, Debug, PartialEq)]
pub enum CounterInstruction {
    /// Increments the counter by 1
    ///
    /// Accounts expected:
    /// 0. `[writable]` Account that holds counter state.
    Increment,

    /// Increments the counter by user supplied amount.
    ///
    /// Accounts expected:
    /// 0. `[writable]` Account that holds counter state.
    IncrementBy { amount: u8 },

    /// Decrements the counter by 1
    ///
    /// Accounts expected:
    /// 0. `[writable]` Account that holds counter state.
    Decrement,

    /// Decrements the counter by user supplied amount.
    ///
    /// Accounts expected:
    /// 0. `[writable]` Account that holds counter state.
    DecrementBy { amount: u8 },

    /// Resets the counter to zero.
    ///
    /// Accounts expected:
    /// 0. `[writable]` Account that holds counter state.
    Reset,

    /// Resets the counter to to the supplied value.
    ///
    /// Accounts expected:
    /// 0. `[writable]` Account that holds counter state.
    ResetTo { value: u16 },
}

#[cfg(test)]
mod tests {
    use crate::instruction::CounterInstruction;
    use borsh::{BorshDeserialize, BorshSerialize};

    #[test]
    fn test_increment_serialization() {
        let original = CounterInstruction::Increment;
        let encoded = original.try_to_vec().expect("Could not encode");
        let decoded = CounterInstruction::try_from_slice(&encoded).expect("Could not decode");
        assert_eq!(original, decoded);
    }

    #[test]
    fn test_increment_by_serialization() {
        let original = CounterInstruction::IncrementBy { amount: 255 };
        let encoded = original.try_to_vec().expect("Could not encode");
        let decoded = CounterInstruction::try_from_slice(&encoded).expect("Could not decode");
        assert_eq!(original, decoded);

        let ix = CounterInstruction::try_from_slice(&[1, 255]).expect("Could not decode");
        assert_eq!(original, ix);
    }

    #[test]
    fn test_decrement_serialization() {
        let original = CounterInstruction::Decrement;
        let encoded = original.try_to_vec().expect("Could not encode");
        let decoded = CounterInstruction::try_from_slice(&encoded).expect("Could not decode");
        assert_eq!(original, decoded);
    }

    #[test]
    fn test_decrement_by_serialization() {
        let original = CounterInstruction::DecrementBy { amount: 5 };
        let encoded = original.try_to_vec().expect("Could not encode");
        let decoded = CounterInstruction::try_from_slice(&encoded).expect("Could not decode");
        assert_eq!(original, decoded);
    }

    #[test]
    fn test_reset_serialization() {
        let original = CounterInstruction::Reset;
        let encoded = original.try_to_vec().expect("Could not encode");
        let decoded = CounterInstruction::try_from_slice(&encoded).expect("Could not decode");
        assert_eq!(original, decoded);
    }

    #[test]
    fn test_reset_to_serialization() {
        let original = CounterInstruction::ResetTo { value: 9000 };
        let encoded = original.try_to_vec().expect("Could not encode");
        let decoded = CounterInstruction::try_from_slice(&encoded).expect("Could not decode");
        assert_eq!(original, decoded);
    }
}

File: 001-counter/client/src/instruction.ts
import { Enum, SOLANA_SCHEMA, Struct } from "@solana/web3.js";

class Increment extends Struct { }

class IncrementBy extends Struct { }

class Decrement extends Struct { }

class DecrementBy extends Struct { }

class Reset extends Struct { }

class ResetTo extends Struct { }

export class CounterInstruction extends Enum {
    static increment(): CounterInstruction {
        return new CounterInstruction({
            increment: new Increment({})
        })
    }
    static incrementBy(amount: number): CounterInstruction {
        return new CounterInstruction({
            incrementBy: new IncrementBy({ amount })
        })
    }
    static decrement(): CounterInstruction {
        return new CounterInstruction({
            decrement: new Decrement({})
        })
    }
    static decrementBy(amount: number): CounterInstruction {
        return new CounterInstruction({
            decrementBy: new DecrementBy({ amount })
        })
    }
    static reset(): CounterInstruction {
        return new CounterInstruction({
            reset: new Reset({})
        })
    }
    static resetTo(value: number): CounterInstruction {
        return new CounterInstruction({
            resetTo: new ResetTo({ value })
        })
    }
}

SOLANA_SCHEMA.set(Increment, {
    kind: 'struct',
    fields: [],
});

SOLANA_SCHEMA.set(IncrementBy, {
    kind: 'struct',
    fields: [
        ['amount', 'u8'],
    ],
});

SOLANA_SCHEMA.set(Decrement, {
    kind: 'struct',
    fields: [],
});

SOLANA_SCHEMA.set(DecrementBy, {
    kind: 'struct',
    fields: [
        ['amount', 'u8'],
    ],
});

SOLANA_SCHEMA.set(Reset, {
    kind: 'struct',
    fields: [],
});

SOLANA_SCHEMA.set(ResetTo, {
    kind: 'struct',
    fields: [
        ['value', 'u16'],
    ],
});

SOLANA_SCHEMA.set(CounterInstruction, {
    kind: 'enum',
    field: 'enum',
    values: [
        ['increment', Increment],
        ['incrementBy', IncrementBy],
        ['decrement', Decrement],
        ['decrementBy', DecrementBy],
        ['reset', Reset],
        ['resetTo', ResetTo],
    ],
});